### PR TITLE
Some refactors of configuration file discovery

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Sorry for this being a bit all over the place...

These are some basic refactors all related to `read_config_file` to make the parameters that get passed to that function a little more logical. That makes it easier to write a new function that doesn't need to handle some of the non-sensical parameters.

I also removed the configuration file discovery out of `PyLinter` into `Run`. Since the configuration is done in `Run` it makes no sense to do that discovery in `PyLinter`.

### CODE REVIEW

**Comment by DanielNoord** [comment]
Initial test failure was due to incomplete call to `_config_initialization`. However, since that is private it shouldn't matter too much.

File: pylint/config/option_manager_mixin.py [hunk 1]
```diff
@@ -245,25 +244,23 @@ def load_provider_defaults(self):
...
-        if config_file is None:
-            config_file = self.config_file
-        if config_file is not None:
-            config_file = os.path.expandvars(os.path.expanduser(config_file))
-            if not os.path.exists(config_file):
-                raise OSError(f"The config file {config_file} doesn't exist!")
-
-        use_config_file = config_file and os.path.exists(config_file)
-        if use_config_file:
+        if config_file:
```
Comment (Pierre-Sassoulas) [inline pylint/config/option_manager_mixin.py:253]:
```suggestion
        if not config_file:
            if verbose:
                msg = "No config file found, using default configuration"
                print(msg, file=sys.stderr)
            return
       ...
```
Pre-ma-ture re-turn 🎉  This one actually made you buffer 30 lines and 5+ ifs/else that even github hide as it's too long to show

File: pylint/config/config_initialization.py [hunk 2]
```diff
@@ -18,18 +18,23 @@ def _config_initialization(
...
     config_file: Union[None, str, Path] = None,
-    verbose_mode: Optional[bool] = None,
+    verbose_mode: bool = False,
 ) -> List[str]:
     """Parse all available options, read config files and command line arguments and
     set options accordingly.
     """
+    # Convert config_file in a Path and string
+    config_file_path = Path(config_file) if config_file else None
+    config_file = str(config_file) if config_file else None
```
Comment (Pierre-Sassoulas) [inline pylint/config/config_initialization.py:28]: Why do we need both ?  
```suggestion
    config_file = Path(config_file) if config_file else None
```
The we cast to string when/if required ?

File: pylint/config/config_initialization.py [hunk 3]
```diff
@@ -18,18 +18,23 @@ def _config_initialization(
...
+
     # Set the current module to the configuration file
     # to allow raising messages on the configuration file.
-    linter.set_current_module(linter.config_file)
+    linter.set_current_module(config_file)
```
Comment (Pierre-Sassoulas) [inline pylint/config/config_initialization.py:32]:
```suggestion
    linter.set_current_module(str(config_file))
```

Or we could make ``set_current_module`` accepts Path too.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
I started to modify with the online editor but this is risky as it's impossible to launch tests 😄

File: pylint/config/config_initialization.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/config/config_initialization.py:28]: I opted to do it like this to give some flexibility in later PRs and keep those diffs smaller.

File: pylint/config/option_manager_mixin.py (hunk 1 above)
Comment (DanielNoord) [reply pylint/config/option_manager_mixin.py:253]: I do this in the final version but didn't want to touch the rest of the code too much. It's more about accepting different types of parameters correctly than actually improving this method.

File: pylint/config/option_manager_mixin.py [hunk 4]
```diff
@@ -274,17 +271,16 @@ def read_config_file(self, config_file=None, verbose=None):
                         sect = sect[len("pylint.") :]
                     if not sect.isupper() and values:
                         parser._sections[sect.upper()] = values
+
         if not verbose:
             return
-        if use_config_file:
-            msg = f"Using config file {os.path.abspath(config_file)}"
+        if config_file and config_file.exists():
+            msg = f"Using config file {str(config_file)}"
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylint/config/option_manager_mixin.py:277]: Do `Paths` automatically get converted to string in f-strings?

File: pylint/config/config_initialization.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pylint/config/config_initialization.py:28]: All right !

File: pylint/config/option_manager_mixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py:278]: If we don't change the initialization it's already a string and not a Path, but yes:
```python
>>> from pathlib import Path
>>> f"{Path('tests').resolve()}"
'/home/pierre/pylint/tests'

[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py:278]: Everything get cast to string in f-strings :)

File: pylint/config/option_manager_mixin.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py:253]: Fair enough !

File: pylint/config/option_manager_mixin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylint/config/option_manager_mixin.py]: Yeah I wondered if it was `'/home/pierre/pylint/tests'` or `Path('/home/pierre/pylint/tests')`. I think the `repr()` has the `Path()` in it.

[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py]: Ha yes I did the test because of that too. We probably printed lists of Paths before

**Comment by Pierre-Sassoulas** [comment]
Sorry for butchering the history with my edits and revert.

**Pull request merged**